spark/v3.4/spark/src/test/java/org/apache/iceberg/spark/source/TestFilteredWithSystemFunctionScan.java

```diff
@@ -0,0 +1,825 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.spark.source;
+
+import org.apache.iceberg.Table;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.spark.Spark3Util;
+import org.apache.iceberg.spark.SparkTestBaseWithCatalog;
+import org.apache.iceberg.spark.functions.BucketFunction;
+import org.apache.iceberg.spark.functions.DaysFunction;
+import org.apache.iceberg.spark.functions.HoursFunction;
+import org.apache.iceberg.spark.functions.MonthsFunction;
+import org.apache.iceberg.spark.functions.TruncateFunction;
+import org.apache.iceberg.spark.functions.YearsFunction;
+import org.apache.iceberg.util.DateTimeUtil;
+import org.apache.spark.sql.connector.expressions.Expression;
+import org.apache.spark.sql.connector.expressions.FieldReference;
+import org.apache.spark.sql.connector.expressions.LiteralValue;
+import org.apache.spark.sql.connector.expressions.UserDefinedScalarFunc;
+import org.apache.spark.sql.connector.expressions.filter.And;
+import org.apache.spark.sql.connector.expressions.filter.Not;
+import org.apache.spark.sql.connector.expressions.filter.Or;
+import org.apache.spark.sql.connector.expressions.filter.Predicate;
+import org.apache.spark.sql.connector.read.Batch;
+import org.apache.spark.sql.connector.read.ScanBuilder;
+import org.apache.spark.sql.connector.read.SupportsPushDownV2Filters;
+import org.apache.spark.sql.types.DataTypes;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+import org.assertj.core.api.Assertions;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestFilteredWithSystemFunctionScan extends SparkTestBaseWithCatalog {
+
+  @Before
+  public void useCatalog() {
+    sql("USE %s", catalogName);
+  }
+
+  @After
+  public void dropTable() {
+    sql("DROP TABLE IF EXISTS %s PURGE", tableName);
+  }
+
+  @Test
+  public void testUnpartitionedYears() throws Exception {
+    createUnpartitionedTable();
+
+    SparkScanBuilder builder = scanBuilder();
+
+    YearsFunction.TimestampToYearsFunction function = new YearsFunction.TimestampToYearsFunction();
+    UserDefinedScalarFunc udf =
+        new UserDefinedScalarFunc(
+            function.name(), function.canonicalName(), expressions(FieldReference.apply("ts")));
+    Predicate predicate =
+        new Predicate(
+            "=", expressions(udf, LiteralValue.apply(years("2017-11-22"), DataTypes.IntegerType)));
+    pushFilters(builder, predicate);
+    Batch scan = builder.build().toBatch();
+
+    Assertions.assertThat(scan.planInputPartitions().length).isEqualTo(10);
+
+    // NOT Equal
+    builder = scanBuilder();
+
+    predicate = new Not(predicate);
+    pushFilters(builder, predicate);
+    scan = builder.build().toBatch();
+
+    Assertions.assertThat(scan.planInputPartitions().length).isEqualTo(10);
```

Comment:
Are there any tests that exercise the entire code path using SQL? I'd like to see at least one end-to-end test with SQL.